demo_nodes_cpp/src/events/matched_event_detect.cpp

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 Sony Group Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcl/event.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_components/register_node_macro.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "demo_nodes_cpp/visibility_control.h"
+
+namespace demo_nodes_cpp
+{
+
+// This demo program shows detected matched event.
+// Matched event occur while connection between publisher and subscription is done.
+// Run this demo at one console by below command
+// $ ros2 run demo_nodes_cpp matched_event_detect
+// It will create 2 topics.
+// - 'pub_matched_event_detect' for detecting matched event of publisher
+// - 'sub_matched_event_detect' for detecting matched event of subscription
+//
+// For checking matched event of publisher
+// On another console, run below command
+// $ ros2 topic echo /pub_matched_event_detect
+// You can run above command many times on different consoles to check the number of connected
+// subscription by the output of demo program.
+//
+// For checking matched event of subscription
+// On another console, run below command
+// ros2 topic pub -r 1 /sub_matched_event_detect std_msgs/String "{data: '123'}"
+// You can run above command many times on different consoles to check the number of connected
+// publisher by the output of demo program.
+
+class MatchedEventDetectNode : public rclcpp::Node
+{
+public:
+  DEMO_NODES_CPP_PUBLIC explicit MatchedEventDetectNode(const rclcpp::NodeOptions & options)
+  : Node("matched_event_detection_node", options)
+  {
+    rclcpp::PublisherOptions pub_options;
+    pub_options.event_callbacks.matched_callback =
+      [this](rmw_matched_status_t & s) {
```

Comment:
use `rclcpp::MatchedInfo` instead?

Reply:
> a new test is required in https://github.com/ros2/demos/tree/9fbd3e840113e183b7ccc935e17fa275fdea6540/demo_nodes_cpp/test

`matched_event_detect` like 'serive'. What output depend on 'client' behavior.  
This PR doesn't provide 'client'. 
I want to use 'ros2 topic echo' and 'ros topic pub' command as 'client'.   
Not sure whether current test framework can support it or not.